File: step_definitions/Verify availability of real-time shipping quotes during checkout.js
import { Given, When, Then } from "cypress-cucumber-preprocessor/steps";
import CheckoutPage from "../pages/checkoutPage";
import LoginPage from "../pages/loginPage";
import CartPage from "../pages/cartPage";
import OrderHistoryPage from "../pages/orderHistoryPage";

Given("a user account exists", () => {
 // Assume user account setup is done
});

When("the user logs into the application using valid credentials", () => {
 LoginPage.visit();
 LoginPage.enterCredentials();
 LoginPage.submit();
});

Then("the user is successfully logged in and navigated to the homepage", () => {
 cy.url().should("include", "/homepage");
});

Given("the user is logged in", () => {
 // Assume user is logged in
});

When("the user adds a product to the cart", () => {
 CartPage.addProductToCart();
});

Then("all items added to the cart are displayed with their respective details", () => {
 CartPage.verifyItemsInCart();
});

Given("the user is on the shopping cart page", () => {
 CartPage.visit();
});

When("the user proceeds to checkout", () => {
 CartPage.proceedToCheckout();
});

Then("the checkout page is displayed with options to enter shipping information", () => {
 CheckoutPage.verifyCheckoutPage();
});

Given("the user is on the checkout page", () => {
 CheckoutPage.visit();
});

When("the user enters valid shipping address details", () => {
 CheckoutPage.enterShippingAddress();
});

Then("the shipping address is accepted and saved", () => {
 CheckoutPage.verifyShippingAddressSaved();
});

Given("the user has entered a valid shipping address", () => {
 // Assume shipping address is entered
});

When("the user selects 'Get Shipping Quotes' option", () => {
 CheckoutPage.getShippingQuotes();
});

Then("real-time shipping quotes are displayed from available providers", () => {
 CheckoutPage.verifyShippingQuotesDisplayed();
});

Given("real-time shipping quotes are displayed", () => {
 // Assume shipping quotes are displayed
});

When("the user reviews the shipping quotes and selects a preferred provider", () => {
 CheckoutPage.selectPreferredShippingProvider();
});

Then("the selected shipping provider is highlighted and confirmed for checkout", () => {
 CheckoutPage.verifySelectedProviderConfirmed();
});

Given("the user has selected a shipping provider", () => {
 // Assume shipping provider is selected
});

When("the user confirms the shipping selection and proceeds to payment", () => {
 CheckoutPage.confirmShippingSelection();
 CheckoutPage.proceedToPayment();
});

Then("the user is navigated to the payment section with selected shipping details", () => {
 CheckoutPage.verifyPaymentSectionNavigation();
});

Given("the user is on the payment section", () => {
 CheckoutPage.visitPaymentSection();
});

When("the user enters payment information and completes the transaction", () => {
 CheckoutPage.enterPaymentInformation();
 CheckoutPage.completeTransaction();
});

Then("the transaction is successful, and order confirmation is displayed", () => {
 CheckoutPage.verifyTransactionSuccess();
});

Given("the transaction is successful", () => {
 // Assume transaction is successful
});

When("the user reviews the order summary", () => {
 CheckoutPage.reviewOrderSummary();
});

Then("the order summary includes the selected shipping provider and cost", () => {
 CheckoutPage.verifyOrderSummaryDetails();
});

Given("the order is confirmed", () => {
 // Assume order is confirmed
});

When("the user checks their email", () => {
 CheckoutPage.checkEmailForConfirmation();
});

Then("the email confirmation includes selected shipping provider and estimated delivery date", () => {
 CheckoutPage.verifyEmailConfirmationDetails();
});

Given("the user is logged in", () => {
 // Assume user is logged in
});

When("the user logs out from the application", () => {
 LoginPage.logout();
});

Then("the user is successfully logged out", () => {
 cy.url().should("include", "/login");
});

Given("the user is logged out", () => {
 // Assume user is logged out
});

When("the user re-logs into the application and navigates to order history", () => {
 LoginPage.reLogin();
 OrderHistoryPage.visit();
});

Then("the order history displays recent order with correct shipping details", () => {
 OrderHistoryPage.verifyRecentOrderDetails();
});

Given("the user is in the order history section", () => {
 OrderHistoryPage.visit();
});

When("the user selects the order to view detailed shipping information", () => {
 OrderHistoryPage.viewOrderDetails();
});

Then("detailed shipping information including provider and tracking number is displayed", () => {
 OrderHistoryPage.verifyShippingInformation();
});

Given("the user has a tracking number", () => {
 // Assume tracking number is available
});

When("the user tracks the order using the provided tracking number", () => {
 OrderHistoryPage.trackOrder();
});

Then("tracking information is accessible and updates are displayed in real-time", () => {
 OrderHistoryPage.verifyTrackingUpdates();
});

Given("the user has tracked the order", () => {
 // Assume order is tracked
});

When("the user verifies the estimated delivery date", () => {
 OrderHistoryPage.verifyEstimatedDeliveryDate();
});

Then("the estimated delivery date matches the shipping provider's quote", () => {
 OrderHistoryPage.verifyDeliveryDateMatchesQuote();
});

File: step_definitions/Verify availability of real-time shipping quotes during checkout.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import com.page_objects.CheckoutPage;

public class CheckoutSteps extends CheckoutPage {

    private TestHarness testHarness = new TestHarness();

    @Given("a user account exists")
    public void aUserAccountExists() {
        // Logic to ensure user account exists
    }

    @When("the user logs into the application using valid credentials")
    public void theUserLogsIntoTheApplicationUsingValidCredentials() {
        String userName = testHarness.getData("LoginData", "UserName");
        String password = testHarness.getData("LoginData", "Password");
        enterUserName(userName);
        enterPassword(password);
        clickLoginButton();
    }

    @Then("the user is successfully logged in and navigated to the homepage")
    public void theUserIsSuccessfullyLoggedInAndNavigatedToTheHomepage() {
        String expectedURL = testHarness.getData("LoginData", "ExpectedHomePageURL");
        verifyHomePageRedirection(expectedURL);
    }

    @Given("the user is logged in")
    public void theUserIsLoggedIn() {
        // Logic to ensure user is logged in
    }

    @When("the user adds a product to the cart")
    public void theUserAddsAProductToTheCart() {
        String productName = testHarness.getData("CartData", "ProductName");
        String quantity = testHarness.getData("CartData", "Quantity");
        addProductToCart(productName, quantity);
    }

    @Then("all items added to the cart are displayed with their respective details")
    public void allItemsAddedToTheCartAreDisplayedWithTheirRespectiveDetails() {
        // Logic to verify items in the cart
    }

    @Given("the user is on the shopping cart page")
    public void theUserIsOnTheShoppingCartPage() {
        // Logic to ensure user is on the shopping cart page
    }

    @When("the user proceeds to checkout")
    public void theUserProceedsToCheckout() {
        proceedToCheckout();
    }

    @Then("the checkout page is displayed with options to enter shipping information")
    public void theCheckoutPageIsDisplayedWithOptionsToEnterShippingInformation() {
        // Logic to verify checkout page
    }

    @Given("the user is on the checkout page")
    public void theUserIsOnTheCheckoutPage() {
        // Logic to ensure user is on the checkout page
    }

    @When("the user enters valid shipping address details")
    public void theUserEntersValidShippingAddressDetails() {
        String fullName = testHarness.getData("CheckoutData", "FullName");
        String address = testHarness.getData("CheckoutData", "Address");
        String city = testHarness.getData("CheckoutData", "City");
        String postalCode = testHarness.getData("CheckoutData", "PostalCode");
        String phone = testHarness.getData("CheckoutData", "Phone");
        enterShippingDetails(fullName, address, city, postalCode, phone);
    }

    @Then("the shipping address is accepted and saved")
    public void theShippingAddressIsAcceptedAndSaved() {
        // Logic to verify shipping address is saved
    }

    @Given("the user has entered a valid shipping address")
    public void theUserHasEnteredAValidShippingAddress() {
        // Logic to ensure valid shipping address is entered
    }

    @When("the user selects 'Get Shipping Quotes' option")
    public void theUserSelectsGetShippingQuotesOption() {
        // Logic to get shipping quotes
    }

    @Then("real-time shipping quotes are displayed from available providers")
    public void realTimeShippingQuotesAreDisplayedFromAvailableProviders() {
        // Logic to verify shipping quotes
    }

    @Given("real-time shipping quotes are displayed")
    public void realTimeShippingQuotesAreDisplayed() {
        // Logic to ensure shipping quotes are displayed
    }

    @When("the user reviews the shipping quotes and selects a preferred provider")
    public void theUserReviewsTheShippingQuotesAndSelectsAPreferredProvider() {
        // Logic to select preferred shipping provider
    }

    @Then("the selected shipping provider is highlighted and confirmed for checkout")
    public void theSelectedShippingProviderIsHighlightedAndConfirmedForCheckout() {
        // Logic to confirm selected shipping provider
    }

    @Given("the user has selected a shipping provider")
    public void theUserHasSelectedAShippingProvider() {
        // Logic to ensure shipping provider is selected
    }

    @When("the user confirms the shipping selection and proceeds to payment")
    public void theUserConfirmsTheShippingSelectionAndProceedsToPayment() {
        // Logic to proceed to payment
    }

    @Then("the user is navigated to the payment section with selected shipping details")
    public void theUserIsNavigatedToThePaymentSectionWithSelectedShippingDetails() {
        // Logic to verify navigation to payment section
    }

    @Given("the user is on the payment section")
    public void theUserIsOnThePaymentSection() {
        // Logic to ensure user is on payment section
    }

    @When("the user enters payment information and completes the transaction")
    public void theUserEntersPaymentInformationAndCompletesTheTransaction() {
        // Logic to complete transaction
    }

    @Then("the transaction is successful, and order confirmation is displayed")
    public void theTransactionIsSuccessfulAndOrderConfirmationIsDisplayed() {
        // Logic to verify order confirmation
    }

    @Given("the transaction is successful")
    public void theTransactionIsSuccessful() {
        // Logic to ensure transaction is successful
    }

    @When("the user reviews the order summary")
    public void theUserReviewsTheOrderSummary() {
        // Logic to review order summary
    }

    @Then("the order summary includes the selected shipping provider and cost")
    public void theOrderSummaryIncludesTheSelectedShippingProviderAndCost() {
        // Logic to verify order summary
    }

    @Given("the order is confirmed")
    public void theOrderIsConfirmed() {
        // Logic to ensure order is confirmed
    }

    @When("the user checks their email")
    public void theUserChecksTheirEmail() {
        // Logic to check email
    }

    @Then("the email confirmation includes selected shipping provider and estimated delivery date")
    public void theEmailConfirmationIncludesSelectedShippingProviderAndEstimatedDeliveryDate() {
        // Logic to verify email confirmation
    }

    @Given("the user is logged in")
    public void theUserIsLoggedInAgain() {
        // Logic to ensure user is logged in
    }

    @When("the user logs out from the application")
    public void theUserLogsOutFromTheApplication() {
        // Logic to log out
    }

    @Then("the user is successfully logged out")
    public void theUserIsSuccessfullyLoggedOut() {
        // Logic to verify logout
    }

    @Given("the user is logged out")
    public void theUserIsLoggedOut() {
        // Logic to ensure user is logged out
    }

    @When("the user re-logs into the application and navigates to order history")
    public void theUserReLogsIntoTheApplicationAndNavigatesToOrderHistory() {
        // Logic to re-log in and navigate to order history
    }

    @Then("the order history displays recent order with correct shipping details")
    public void theOrderHistoryDisplaysRecentOrderWithCorrectShippingDetails() {
        // Logic to verify order history
    }

    @Given("the user is in the order history section")
    public void theUserIsInTheOrderHistorySection() {
        // Logic to ensure user is in order history section
    }

    @When("the user selects the order to view detailed shipping information")
    public void theUserSelectsTheOrderToViewDetailedShippingInformation() {
        // Logic to view detailed shipping information
    }

    @Then("detailed shipping information including provider and tracking number is displayed")
    public void detailedShippingInformationIncludingProviderAndTrackingNumberIsDisplayed() {
        // Logic to verify detailed shipping information
    }

    @Given("the user has a tracking number")
    public void theUserHasATrackingNumber() {
        // Logic to ensure tracking number is available
    }

    @When("the user tracks the order using the provided tracking number")
    public void theUserTracksTheOrderUsingTheProvidedTrackingNumber() {
        // Logic to track order
    }

    @Then("tracking information is accessible and updates are displayed in real-time")
    public void trackingInformationIsAccessibleAndUpdatesAreDisplayedInRealTime() {
        // Logic to verify tracking information
    }

    @Given("the user has tracked the order")
    public void theUserHasTrackedTheOrder() {
        // Logic to ensure order is tracked
    }

    @When("the user verifies the estimated delivery date")
    public void theUserVerifiesTheEstimatedDeliveryDate() {
        // Logic to verify estimated delivery date
    }

    @Then("the estimated delivery date matches the shipping provider's quote")
    public void theEstimatedDeliveryDateMatchesTheShippingProvidersQuote() {
        // Logic to verify estimated delivery date
    }
}